yapapi/storage/gftp.py

```diff
@@ -197,21 +216,58 @@ async def download_file(self, destination_file: PathLike):
         return await super().download_file(destination_file)
 
 
+def _delete_if_exists(path: Path) -> None:
+
+    if path.exists():
+        path.unlink()
+        _logger.debug("Deleted temporary file %s", path)
+
+
 class GftpProvider(StorageProvider, AsyncContextManager[StorageProvider]):
-    _temp_dir: Optional[Path]
-    _registered_sources: Dict[str, GftpSource]
+    """A StorageProvider that communicates with `gftp server` through JSON-RPC."""
+
+    @dataclass
+    class URLInfo:
+        """Information about an URL published through `gftp`."""
+
+        publish_count: int
+        """Number of `gftp publish` operations for this URL.
+
+        Serves as a reference counter. When it drops to 0, `gftp close {URL}` is invoked
+        in order to release any file published with this URL that is kept open by `gftp`.
+        Note that the value of this field may be larger than the number of files published,
+        since a single file may be published more than once."""
+
+        temporary_files: Set[Path]
+        """Set of temporary files published with this URL.
 
-    def __init__(self, *, tmpdir: Optional[str] = None):
+        When the URL is unpublished by calling `gftp close {URL}`, all temporary files with this
+        URL can be safely deleted.
+        """
+
+    def __init__(self, *, tmpdir: Optional[str] = None, _close_urls: bool = False):
```

Comment:
GitHub doesn't let me make a suggestion on a line range which includes deleted lines, so let me paste it here:
```
    _close_urls: bool
    """Flag indicating if this `GftpProvider` will close unpublished URLs.

    If set to `True` then the provider will call `gftp close <URL>` for any URL
    that no longer has any published files. This should be the default behavior,
    but is currently turned off until `gftp` is able to handle the `close` command
    correctly (see https://github.com/golemfactory/yagna/pull/1501)."""

    _lock: asyncio.Lock
    """Lock used to synchronize access to `self._published_sources`."""

    _process: "Optional[__Process]"
    """Reference to an external process running the `gftp server` command."""

    _published_sources: "Dict[str, GftpProvider.URLInfo]"
    """Mapping of URLs to info on files published with the given URL."""

    _temp_dir: Optional[Path]

    def __init__(self, *, tmpdir: Optional[str] = None, _close_urls: bool = False):
        self.__exit_stack = AsyncExitStack()
        self._close_urls = _close_urls
        self._lock = asyncio.Lock()
        self._process = None
        self._published_sources: Dict[str, GftpProvider.URLInfo] = dict()
        self._temp_dir: Optional[Path] = Path(tmpdir) if tmpdir else None
```

Reply:
I'm not sure we should add docstrings for non-public attributes. I know we do it sometimes in our new code, but since I don't have strong opinion on this, I'd prefer to leave it as it was in the original code. I've added type annotations though in 647a4b0